ports/samd/modutime.c

```diff
@@ -24,39 +24,26 @@
  * THE SOFTWARE.
  */
 
-#include "py/runtime.h"
-#include "extmod/utime_mphal.h"
+#include "py/obj.h"
 #include "shared/timeutils/timeutils.h"
 #include "mphalport.h"
 
 #if !MICROPY_PY_MACHINE_RTC
 uint32_t time_offset = 0;
 #endif // !MICROPY_PY_MACHINE_RTC
 
-// localtime([secs])
-STATIC mp_obj_t time_localtime(size_t n_args, const mp_obj_t *args) {
+// Return the localtime as an 8-tuple.
+STATIC mp_obj_t mp_utime_localtime_get(void) {
     timeutils_struct_time_t tm;
-    mp_int_t seconds;
 
     #if MICROPY_PY_MACHINE_RTC
     extern void rtc_gettime(timeutils_struct_time_t *tm);
-    if (n_args == 0 || args[0] == mp_const_none) {
-        rtc_gettime(&tm);
-    } else {
-        seconds = mp_obj_get_int(args[0]);
-        timeutils_seconds_since_epoch_to_struct_time(seconds, &tm);
-    }
-
+    rtc_gettime(&tm);
     #else
-    if (n_args == 0 || args[0] == mp_const_none) {
-        seconds = mp_hal_ticks_ms_64() / 1000 + time_offset;
-    } else {
-        seconds = mp_obj_get_int(args[0]);
-        time_offset = seconds - mp_hal_ticks_ms_64() / 1000;
```

Comment:
@robert-hh there is a problem here: `time_offset` is no longer set anywhere.

What was the original idea with `time_offset`?  It seems that calling `time.localtime(123...)` would set this offset, but I don't understand the logic behind that.

Reply:
The intention was a means to set the actual time for boards without the RTC module. The RTC module is not included for boards that have no crystal, like the Adafruit ItsyBitsy ones. Then localtime(123,,,) would set the time, reported later.

Reply:
`time.localtime(123...)` should return a struct/tuple containing the time in seconds (that was passed in) expressed as year/month/day/hour/minute/second/etc.  It's definitely unexpected behaviour for that to *set* the time (and I do now see that in the quickref, now that you mention it).

It's going to be awkward to have this a special case just for samd.  So let's think of another way to set the time on devices without RTC.  Maybe without an RTC it still has the `machine.RTC()` class, but it can only be used to set the time?

Reply:
I agree that this special method is kind of unusual. For SAMD, it could always enable RTC,  Only the time precision would be bad, since it would use the internal ULP RC-Oscillator as time reference. And then people would complain. At the moment without RTC, localtime uses the systick as reference, which, when connected to USB, is pretty precise. If not, it falls back to the calibrated 48Mhz clock, which is still better than the RC-Oscillator.

How do other portss without the RTC module set the time/date for localtime?

Reply:
> How do other portss without the RTC module set the time/date for localtime?

All of the main ports have `machine.RTC()`, from what I can tell.

`nrf` does not, but it also doesn't implement `time.localtime()` or `time.time()`.

Reply:
Then it may be better to enable RTC as well, even if the quality is poor without a crystal. Would you change that as part for this PR, or should I make a change for it? It affects the mpconfigmcu.h files and would be set in mpconfigport.h.
Correction to my statement above: If there is no crystal, RTC is always operated by the low power RC oscillator, which is quite poor in matters of frequency match and temperature changes.

Reply:
Yes I think it's best to enable the RTC.  That keeps consistency across ports.

The function `time.ticks_ms()` and related ones still use the systick (right?), so can be more precise than the RTC.

If you can make this change separately, that would be great.

Reply:
The function time.ticks_ms() use the µs counter, which is driven by the 48MHz clock, which is controlled by crystal, USB or internal calibration, whichever is available. Systick runs from the CPU clock, which on the SAMD21 use either  the ULP32K RC oscillator or the crystal as reference. On the SAMD51 the CPU clock reference is derived from the 48MHz clock or ULP32K RC oscillator, if the respective flag is set. There are extensive comments in clock_config.c about that, for me to remember. So systick is only used for the scheduler.
And yes, I'll make the change.

Reply:
Time keeping is hard to get right!

Reply:
Addressed by #10996.